src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -321,6 +321,20 @@ internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree
                 // Consider reporting the correct source with annotations on mismatch.
                 AssertEx.Equal(expectedTypes, actualTypes, message: method.ToTestDisplayString());
 
+                foreach (var entry in dictionary.Values.Where(v => !v.IsNull))
+                {
+                    // Result types cannot have nested types that are lossy unspeakables
+                    Assert.Null(entry.VisitType(typeOpt: null,
+                        typeWithAnnotationsPredicateOpt: (tswa, a, b) => !tswa.Equals(entry, TypeCompareKind.ConsiderEverything) && isLossyUnspeakable(tswa),
+                        typePredicateOpt: (ts, _, b) => false, arg: (object)null, canDigThroughNullable: true));
+                }
+
+                // Returns true if the application of TypeSymbolWithAnnotations.AsSpeakable would lose information
+                bool isLossyUnspeakable(TypeSymbolWithAnnotations tswa)
+                {
+                    return tswa.NullableAnnotation == NullableAnnotation.NotNullable && tswa.TypeSymbol.IsTypeParameterDisallowingAnnotation();
```

Comment:
>NullableAnnotation.NotNullable [](start = 54, length = 30)

What about ```NullableAnnotation.Nullable```?

Reply:
If you have a `T` that is `Nullable` and disallows annotation, changing it to `NotAnnotated` (as done by `AsSpeakable()`) does not lose information.

That said, a stronger assertion can be made here, so I'll modify.

---
In reply to: [244061708](https://github.com/dotnet/roslyn/pull/31813#discussion_r244061708) [](ancestors = 244061708)